Add bindings for DiscreteDerivative

danmar3 opened the pull request:
Add bindings for DiscreteDerivative and StateInterpolatorWithDiscreteDerivative.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/12770)
<!-- Reviewable:end -->

jwnimmer-tri commented:
@drake-jenkins-bot ok to test

jwnimmer-tri commented:
Hi @danmar3.  Thanks for your contribution!

For bindings PRs, we typically ask for one small additional task:

> In general, since the Python bindings wrap tested C++ code, you do not (and should not) repeat intricate testing logic done in C++. Instead, ensure you exercise the Pythonic portion of the API, using kwargs when appropriate.

-- https://drake.mit.edu/doxygen_cxx/group__python__bindings.html

Basically we'd like to see a new smoke test case added to https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/systems/test/primitives_test.py that calls each new binding once, as a basic sanity check.

See #12020 for a good example.

danmar3 commented:
Thank you all for taking a look at this. I pushed the changes addressing the reviews. 
I have not used reviewable before, so I'm not sure if there is something else to do.



> <img alt=":lgtm:" width="61" height="20" src="https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67"> feature review -- just one thing to improve, see below.
> 
> +@ggould-tri for platform review per schedule, please.
> 
> > Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
> > _[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/12770#-:-M15cA65Bikhmn4sCKNz:b-dm4kaz)_ status: 2 unresolved discussions, LGTM missing from assignee ggould-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @danmar3 and @ggould-tri)
> 
> _[bindings/pydrake/systems/primitives_py.cc, line 262 at r3](https://reviewable.io/reviews/robotlocomotion/drake/12770#-M15bspo3bDNMAn-xenY:-M15bspo3bDNMAn-xenZ:b-p0nsmq) ([raw file](https://github.com/robotlocomotion/drake/blob/ffe7a236b1a5e2479ed09c6a98c77f306bb00b75/bindings/pydrake/systems/primitives_py.cc#L262)):_
> 
> > ```c++
> >               return self->set_initial_position(context, position);
> >             },
> >             py_reference, py::arg("context"), py::arg("position"),
> > ```
> 
> nit Because this method returns void, the `py_reference,` annotation is irrelevant and should be removed.
> 
> _[bindings/pydrake/systems/primitives_py.cc, line 270 at r3](https://reviewable.io/reviews/robotlocomotion/drake/12770#-M15c4Pe-uJwXSOj2j1S:-M15c4Pe-uJwXSOj2j1T:b-p0nsmq) ([raw file](https://github.com/robotlocomotion/drake/blob/ffe7a236b1a5e2479ed09c6a98c77f306bb00b75/bindings/pydrake/systems/primitives_py.cc#L270)):_
> 
> > ```c++
> >               return self->set_initial_position(state, position);
> >             },
> >             py_reference, py::arg("state"), py::arg("position"),
> > ```
> 
> nit Because this method returns void, the `py_reference,` annotation is irrelevant and should be removed.
